diff -NBaur pspsdk-1.0+beta/tools/psp-config.c pspsdk-1.0+beta-new/tools/psp-config.c
--- pspsdk-1.0+beta/tools/psp-config.c	Mon Jun 27 06:31:22 2005
+++ pspsdk-1.0+beta-new/tools/psp-config.c	Tue Jul  5 04:54:45 2005
@@ -17,8 +19,17 @@
 #define PATH_SEP ":"
 #define DIR_SEP '/'
 #define DIR_SEP_STR "/"
+
+#ifdef __MINGW32__
+
+/* The suffix to the path to strip off, if this is not there then we have an error */
+#define PSPDEV_PATH_SUFFIX "/bin/psp-config.exe"
+#else
+
 /* The suffix to the path to strip off, if this is not there then we have an error */
 #define PSPDEV_PATH_SUFFIX "/bin/psp-config"
+
+#endif
 /************************/
 
 /* Speficies that the current usage is to the print the pspsdk path */
@@ -70,6 +81,20 @@
 	int found = 0;
 
 	/* Check if name is an absolute path, if so out job is done */
+#ifdef __MINGW32__
+
+	char *ptr = name;
+	
+	char temp = name[0];
+	*(ptr++) = '/';
+	*(ptr++) = temp;
+	while (*(ptr)) {
+		temp = *(ptr);
+		if (temp == '\\') *(ptr) = '/';
+		ptr++;
+	}
+
+#endif
 	if(name[0] == DIR_SEP)
 	{
 		/* Absolute path */
