diff -Nbaur binutils-2.18.50/config.sub binutils-2.18.50-ppc/config.sub
--- binutils-2.18.50/config.sub	Wed Aug 29 16:23:21 2007
+++ binutils-2.18.50-ppc/config.sub	Fri Dec 14 17:16:53 2007
@@ -230,6 +230,10 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+	-gekko)
+		basic_machine=powerpc-eabi
+		os=-elf
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
diff -Nbaur binutils-2.18.50/gas/config/tc-ppc.c binutils-2.18.50-ppc/gas/config/tc-ppc.c
--- binutils-2.18.50/gas/config/tc-ppc.c	Mon Nov 12 00:27:53 2007
+++ binutils-2.18.50-ppc/gas/config/tc-ppc.c	Sat Dec 15 12:14:55 2007
@@ -847,6 +847,9 @@
   /* Do all PPC750s have paired single ops?  */
   else if (strcmp (arg, "750cl") == 0)
     ppc_cpu = PPC_OPCODE_PPC | PPC_OPCODE_PPCPS;
+  /* PowerPC Gekko  */
+  else if (strcmp (arg, "gekko") == 0)
+    ppc_cpu = PPC_OPCODE_CLASSIC | PPC_OPCODE_PPC | PPC_OPCODE_32 | PPC_OPCODE_PPCPS;
   /* -m403 and -m405 mean to assemble for the PowerPC 403/405.  */
   else if (strcmp (arg, "403") == 0
 	   || strcmp (arg, "405") == 0)
@@ -1115,7 +1118,8 @@
 -m440			generate code for PowerPC 440\n\
 -m7400, -m7410, -m7450, -m7455\n\
 			generate code for PowerPC 7400/7410/7450/7455\n\
--m750cl			generate code for PowerPC 750cl\n"));
+-m750cl			generate code for PowerPC 750cl\n\
+-mgekko			generate code for PowerPC Gekko\n"));
   fprintf (stream, _("\
 -mppc64, -m620		generate code for PowerPC 620/625/630\n\
 -mppc64bridge		generate code for PowerPC 64, including bridge insns\n\
