diff -Nabur binutils-2.20.1/gas/config/tc-ppc.c binutils-2.20.1-ppc/gas/config/tc-ppc.c
--- binutils-2.20.1/gas/config/tc-ppc.c	2010-02-22 08:06:52.000000000 +0000
+++ binutils-2.20.1-ppc/gas/config/tc-ppc.c	2010-06-17 15:10:26.000000000 +0100
@@ -1193,7 +1193,8 @@
 -m476			generate code for PowerPC 476\n\
 -m7400, -m7410, -m7450, -m7455\n\
 			generate code for PowerPC 7400/7410/7450/7455\n\
--m750cl			generate code for PowerPC 750cl\n"));
+-m750cl			generate code for PowerPC 750cl\n\
+-mgekko			generate code for PowerPC Gekko\n"));
   fprintf (stream, _("\
 -mppc64, -m620		generate code for PowerPC 620/625/630\n\
 -mppc64bridge		generate code for PowerPC 64, including bridge insns\n\
diff -Nabur binutils-2.20.1/opcodes/ppc-dis.c binutils-2.20.1-ppc/opcodes/ppc-dis.c
--- binutils-2.20.1/opcodes/ppc-dis.c	2010-02-22 08:07:12.000000000 +0000
+++ binutils-2.20.1-ppc/opcodes/ppc-dis.c	2010-06-17 15:33:40.000000000 +0100
@@ -89,6 +89,10 @@
     0 },
   { "750cl",   (PPC_OPCODE_PPC | PPC_OPCODE_PPCPS)
     , 0 },
+
+  { "gekko",   (PPC_OPCODE_PPC | PPC_OPCODE_PPCPS | PPC_OPCODE_32 | PPC_OPCODE_CLASSIC)
+    , 0 },
+
   { "a2",      (PPC_OPCODE_PPC | PPC_OPCODE_CLASSIC | PPC_OPCODE_ISEL
 		| PPC_OPCODE_POWER4 | PPC_OPCODE_POWER5 | PPC_OPCODE_CACHELCK
 		| PPC_OPCODE_64 | PPC_OPCODE_A2),
