diff -Nbaur gcc-4.1.0/config.sub gcc-4.1.0-ppc/config.sub
--- gcc-4.1.0/config.sub	Fri Dec 16 12:57:40 2005
+++ gcc-4.1.0-ppc/config.sub	Mon Apr 10 09:48:17 2006
@@ -228,6 +228,10 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+	-gekko)
+		basic_machine=powerpc-eabi
+		os=-elf
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
diff -Nbaur gcc-4.1.0/gcc/c-incpath.c gcc-4.1.0-ppc/gcc/c-incpath.c
--- gcc-4.1.0/gcc/c-incpath.c	Sat Jun 25 03:02:01 2005
+++ gcc-4.1.0-ppc/gcc/c-incpath.c	Mon Apr 10 09:48:17 2006
@@ -331,13 +331,18 @@
   cpp_dir *p;
 
 #if defined (HAVE_DOS_BASED_FILE_SYSTEM)
-  /* Convert all backslashes to slashes.  The native CRT stat()
-     function does not recognize a directory that ends in a backslash
-     (unless it is a drive root dir, such "c:\").  Forward slashes,
-     trailing or otherwise, cause no problems for stat().  */
-  char* c;
-  for (c = path; *c; c++)
-    if (*c == '\\') *c = '/';
+  /* Remove unnecessary trailing slashes.  On some versions of MS
+     Windows, trailing  _forward_ slashes cause no problems for stat().
+     On newer versions, stat() does not recognise a directory that ends
+     in a '\\' or '/', unless it is a drive root dir, such as "c:/",
+     where it is obligatory.  */
+  int pathlen = strlen (path);
+  char* end = path + pathlen - 1;
+  /* Preserve the lead '/' or lead "c:/".  */
+  char* start = path + (pathlen > 2 && path[1] == ':' ? 3 : 1);
+  
+  for (; end > start && IS_DIR_SEPARATOR (*end); end--)
+    *end = 0;
 #endif
 
   p = xmalloc (sizeof (cpp_dir));
diff -Nbaur gcc-4.1.0/gcc/version.c gcc-4.1.0-ppc/gcc/version.c
--- gcc-4.1.0/gcc/version.c	Wed Mar 16 06:04:10 2005
+++ gcc-4.1.0-ppc/gcc/version.c	Mon Apr 10 09:49:51 2006
@@ -8,7 +8,7 @@
    in parentheses.  You may also wish to include a number indicating
    the revision of your modified compiler.  */
 
-#define VERSUFFIX ""
+#define VERSUFFIX " (devkitPPC release 13)"
 
 /* This is the location of the online document giving instructions for
    reporting bugs.  If you distribute a modified version of GCC,
@@ -17,7 +17,7 @@
    forward us bugs reported to you, if you determine that they are
    not bugs in your modifications.)  */
 
-const char bug_report_url[] = "<URL:http://gcc.gnu.org/bugs.html>";
+const char bug_report_url[] = "<URL:http://devkitpro.sourceforge.net/bugs.shtml>";
 
 /* The complete version string, assembled from several pieces.
    BASEVER, DATESTAMP, and DEVPHASE are defined by the Makefile.  */
