diff -Nbaur gdb-7.5.1/gdb/Makefile.in gdb-7.5.1-ppc/gdb/Makefile.in
--- gdb-7.5.1/gdb/Makefile.in	2012-07-02 16:29:33.000000000 +0100
+++ gdb-7.5.1-ppc/gdb/Makefile.in	2012-12-19 01:19:10.000000000 +0000
@@ -2156,10 +2156,10 @@
 # Note that we put the dependencies into a .Tpo file, then move them
 # into place if the compile succeeds.  We need this because gcc does
 # not atomically write the dependency output file.
-@GMAKE_TRUE@override COMPILE.post = -c -o $@ -MT $@ -MMD -MP \
-@GMAKE_TRUE@	-MF $(DEPDIR)/$(basename $(@F)).Tpo
-@GMAKE_TRUE@override POSTCOMPILE = @mv $(DEPDIR)/$(basename $(@F)).Tpo \
-@GMAKE_TRUE@	$(DEPDIR)/$(basename $(@F)).Po
+#@GMAKE_TRUE@override COMPILE.post = -c -o $@ -MT $@ -MMD -MP \
+#@GMAKE_TRUE@	-MF $(DEPDIR)/$(basename $(@F)).Tpo
+#@GMAKE_TRUE@override POSTCOMPILE = @mv $(DEPDIR)/$(basename $(@F)).Tpo \
+#@GMAKE_TRUE@	$(DEPDIR)/$(basename $(@F)).Po
 @GMAKE_TRUE@else
 @GMAKE_TRUE@override COMPILE.pre = source='$<' object='$@' libtool=no \
 @GMAKE_TRUE@	DEPDIR=$(DEPDIR) $(DEPMODE) $(depcomp) $(CC)
diff -Nbaur gdb-7.5.1/sim/ppc/Makefile.in gdb-7.5.1-ppc/sim/ppc/Makefile.in
--- gdb-7.5.1/sim/ppc/Makefile.in	2009-09-16 00:26:44.000000000 +0100
+++ gdb-7.5.1-ppc/sim/ppc/Makefile.in	2012-12-19 01:19:10.000000000 +0000
@@ -551,7 +551,7 @@
 PACKAGE_OBJ = @sim_pk_obj@
 
 
-psim: $(TARGETLIB) main.o $(LIBIBERTY_LIB) $(BFD_LIB) $(LIBS) $(LIBINTL_DEP)
+psim: $(TARGETLIB) main.o $(LIBIBERTY_LIB) $(BFD_LIB) $(LIBINTL_DEP)
 	$(CC) $(CFLAGS) $(SIM_CFLAGS) $(LDFLAGS) -o psim$(EXEEXT) main.o $(TARGETLIB) $(BFD_LIB) $(LIBINTL) $(LIBIBERTY_LIB) $(LIBS)
 
 run: psim
