diff -Nbaur binutils-2.23.1/bfd/configure binutils-2.23.1-arm/bfd/configure
--- binutils-2.23.1/bfd/configure	2012-11-13 14:17:40.000000000 +0000
+++ binutils-2.23.1-arm/bfd/configure	2013-09-04 10:28:02.000000000 +0100
@@ -606,6 +606,7 @@
 tdefaults
 bfd_ufile_ptr
 bfd_file_ptr
+lt_cv_dlopen_libs
 supports_plugins
 bfd_default_target_size
 bfd_machines
diff -Nbaur binutils-2.23.1/bfd/configure.in binutils-2.23.1-arm/bfd/configure.in
--- binutils-2.23.1/bfd/configure.in	2012-11-13 14:17:38.000000000 +0000
+++ binutils-2.23.1-arm/bfd/configure.in	2013-09-04 10:28:02.000000000 +0100
@@ -1098,7 +1098,7 @@
 else
   supports_plugins=0
 fi
-AC_SUBST(supports_plugins)
+AC_SUBST(lt_cv_dlopen_libs)
 
 # Determine the host dependant file_ptr a.k.a. off_t type.  In order
 # prefer: off64_t - if ftello64 and fseeko64, off_t - if ftello and
diff -Nbaur binutils-2.23.1/bfd/Makefile.am binutils-2.23.1-arm/bfd/Makefile.am
--- binutils-2.23.1/bfd/Makefile.am	2012-11-13 14:17:38.000000000 +0000
+++ binutils-2.23.1-arm/bfd/Makefile.am	2013-09-04 10:28:02.000000000 +0100
@@ -35,7 +35,7 @@
 AM_CPPFLAGS = -DBINDIR='"$(bindir)"'
 if PLUGINS
 bfdinclude_HEADERS += $(INCDIR)/plugin-api.h
-LIBDL = -ldl
+LIBDL = @lt_cv_dlopen_libs@
 endif
 
 # bfd.h goes here, for now
diff -Nbaur binutils-2.23.1/bfd/Makefile.in binutils-2.23.1-arm/bfd/Makefile.in
--- binutils-2.23.1/bfd/Makefile.in	2012-11-13 14:17:43.000000000 +0000
+++ binutils-2.23.1-arm/bfd/Makefile.in	2013-09-04 10:28:02.000000000 +0100
@@ -293,6 +293,7 @@
 libexecdir = @libexecdir@
 localedir = @localedir@
 localstatedir = @localstatedir@
+lt_cv_dlopen_libs = @lt_cv_dlopen_libs@
 mandir = @mandir@
 mkdir_p = @mkdir_p@
 oldincludedir = @oldincludedir@
@@ -336,7 +337,7 @@
 @INSTALL_LIBBFD_FALSE@noinst_LTLIBRARIES = libbfd.la
 AM_CFLAGS = $(WARN_CFLAGS)
 AM_CPPFLAGS = -DBINDIR='"$(bindir)"'
-@PLUGINS_TRUE@LIBDL = -ldl
+@PLUGINS_TRUE@LIBDL = @lt_cv_dlopen_libs@
 
 # bfd.h goes here, for now
 BFD_H = bfd.h
diff -Nbaur binutils-2.23.1/include/opcode/arm.h binutils-2.23.1-arm/include/opcode/arm.h
--- binutils-2.23.1/include/opcode/arm.h	2012-09-04 13:53:47.000000000 +0100
+++ binutils-2.23.1-arm/include/opcode/arm.h	2013-09-04 10:28:02.000000000 +0100
@@ -113,7 +113,8 @@
 #define ARM_AEXT_V6KT2  (ARM_AEXT_V6T2 | ARM_EXT_V6K)
 #define ARM_AEXT_V6ZT2  (ARM_AEXT_V6T2 | ARM_EXT_SEC)
 #define ARM_AEXT_V6ZKT2 (ARM_AEXT_V6T2 | ARM_EXT_V6K | ARM_EXT_SEC)
-#define ARM_AEXT_V7_ARM	(ARM_AEXT_V6KT2 | ARM_EXT_V7 | ARM_EXT_BARRIER)
+#define ARM_AEXT_V7_ARM	(ARM_AEXT_V6KT2 | ARM_EXT_V7 | ARM_EXT_BARRIER \
+    | ARM_EXT_OS )
 #define ARM_AEXT_V7A	(ARM_AEXT_V7_ARM | ARM_EXT_V7A)
 #define ARM_AEXT_V7R	(ARM_AEXT_V7_ARM | ARM_EXT_V7R | ARM_EXT_DIV)
 #define ARM_AEXT_NOTM \
