diff -Nbaur binutils-2.17/bfd/doc/chew.c binutils-2.17-arm/bfd/doc/chew.c
--- binutils-2.17/bfd/doc/chew.c	Sun Jul 24 17:57:42 2005
+++ binutils-2.17-arm/bfd/doc/chew.c	Sat Jul  1 18:15:11 2006
@@ -92,6 +92,12 @@
 #define DEF_SIZE 5000
 #define STACK 50

+#ifdef __MINGW32__
+/* Prevent \r\n\ line endings */
+#include <fcntl.h>
+unsigned int _CRT_fmode = _O_BINARY;
+#endif
+
 int internal_wanted;
 int internal_mode;

