diff -Nbaur binutils-2.20/gas/config/tc-arm.c binutils-2.20-arm/gas/config/tc-arm.c
--- binutils-2.20/gas/config/tc-arm.c	2009-08-29 23:10:59.000000000 +0100
+++ binutils-2.20-arm/gas/config/tc-arm.c	2009-11-21 01:29:06.000000000 +0000
@@ -2485,8 +2485,10 @@
       know (frag->tc_frag_data.first_map == NULL);
       frag->tc_frag_data.first_map = symbolP;
     }
+#if ENABLE_CHECKING || defined (DEBUG)
   if (frag->tc_frag_data.last_map != NULL)
     know (S_GET_VALUE (frag->tc_frag_data.last_map) < S_GET_VALUE (symbolP));
+#endif
   frag->tc_frag_data.last_map = symbolP;
 }
 
