diff -Nbaur gdb-7.5.1/gdb/Makefile.in gdb-7.5.1-arm/gdb/Makefile.in
--- gdb-7.5.1/gdb/Makefile.in	2012-07-02 16:29:33.000000000 +0100
+++ gdb-7.5.1-arm/gdb/Makefile.in	2012-12-16 02:41:22.000000000 +0000
@@ -2156,10 +2156,8 @@
 # Note that we put the dependencies into a .Tpo file, then move them
 # into place if the compile succeeds.  We need this because gcc does
 # not atomically write the dependency output file.
-@GMAKE_TRUE@override COMPILE.post = -c -o $@ -MT $@ -MMD -MP \
-@GMAKE_TRUE@	-MF $(DEPDIR)/$(basename $(@F)).Tpo
-@GMAKE_TRUE@override POSTCOMPILE = @mv $(DEPDIR)/$(basename $(@F)).Tpo \
-@GMAKE_TRUE@	$(DEPDIR)/$(basename $(@F)).Po
+@GMAKE_TRUE@override COMPILE.post = -c -o $@
+@GMAKE_TRUE@override POSTCOMPILE =
 @GMAKE_TRUE@else
 @GMAKE_TRUE@override COMPILE.pre = source='$<' object='$@' libtool=no \
 @GMAKE_TRUE@	DEPDIR=$(DEPDIR) $(DEPMODE) $(depcomp) $(CC)
diff -Nbaur gdb-7.5.1/sim/common/Make-common.in gdb-7.5.1-arm/sim/common/Make-common.in
--- gdb-7.5.1/sim/common/Make-common.in	2012-03-19 04:54:47.000000000 +0000
+++ gdb-7.5.1-arm/sim/common/Make-common.in	2012-12-16 02:41:22.000000000 +0000
@@ -417,10 +417,8 @@
 # Note that we put the dependencies into a .Tpo file, then move them
 # into place if the compile succeeds.  We need this because gcc does
 # not atomically write the dependency output file.
-@GMAKE_TRUE@override COMPILE.post = -c -o $@ -MT $@ -MMD -MP \
-@GMAKE_TRUE@	-MF $(DEPDIR)/$(basename $(@F)).Tpo
-@GMAKE_TRUE@override POSTCOMPILE = @mv $(DEPDIR)/$(basename $(@F)).Tpo \
-@GMAKE_TRUE@	$(DEPDIR)/$(basename $(@F)).Po
+@GMAKE_TRUE@override COMPILE.post = -c -o $@
+@GMAKE_TRUE@override POSTCOMPILE =
 @GMAKE_TRUE@else
 @GMAKE_TRUE@override COMPILE.pre = source='$<' object='$@' libtool=no \
 @GMAKE_TRUE@	DEPDIR=$(DEPDIR) $(DEPMODE) $(depcomp) $(CC)
