diff -Nbaur --exclude=*.info* binutils-2.16.1/bfd/doc/chew.c binutils-2.16.1-arm/bfd/doc/chew.c
--- binutils-2.16.1/bfd/doc/chew.c	Thu Mar  3 11:41:02 2005
+++ binutils-2.16.1-arm/bfd/doc/chew.c	Mon Sep 19 03:38:45 2005
@@ -91,6 +91,12 @@
 #define DEF_SIZE 5000
 #define STACK 50
 
+#ifdef __MINGW32__
+/* Prevent \r\n\ line endings */
+#include <fcntl.h>
+unsigned int _CRT_fmode = _O_BINARY;
+#endif
+
 int internal_wanted;
 int internal_mode;
 
diff -Nbaur --exclude=*.info* binutils-2.16.1/configure binutils-2.16.1-arm/configure
--- binutils-2.16.1/configure	Mon Feb 28 14:06:59 2005
+++ binutils-2.16.1-arm/configure	Wed Sep 21 12:33:11 2005
@@ -1034,7 +1034,7 @@
 
 # Some tools are only suitable for building in a "native" situation.
 # Remove these if host!=target.  
-native_only="autoconf automake libtool fileutils find gawk gettext gzip hello indent m4 rcs recode sed shellutils tar textutils uudecode wdiff gprof target-groff guile perl time ash bash bzip2 prms gnuserv target-gperf"
+native_only="autoconf automake libtool fileutils find gawk gettext gzip hello indent m4 rcs recode sed shellutils tar textutils uudecode wdiff  target-groff guile perl time ash bash bzip2 prms gnuserv target-gperf"
 
 # Similarly, some are only suitable for cross toolchains.
 # Remove these if host=target.
diff -Nbaur --exclude=*.info* binutils-2.16.1/configure.in binutils-2.16.1-arm/configure.in
--- binutils-2.16.1/configure.in	Sun Jun 12 20:33:06 2005
+++ binutils-2.16.1-arm/configure.in	Wed Sep 21 12:34:18 2005
@@ -242,7 +242,7 @@
 
 # Some tools are only suitable for building in a "native" situation.
 # Remove these if host!=target.  
-native_only="autoconf automake libtool fileutils find gawk gettext gzip hello indent m4 rcs recode sed shellutils tar textutils uudecode wdiff gprof target-groff guile perl time ash bash bzip2 prms gnuserv target-gperf"
+native_only="autoconf automake libtool fileutils find gawk gettext gzip hello indent m4 rcs recode sed shellutils tar textutils uudecode wdiff target-groff guile perl time ash bash bzip2 prms gnuserv target-gperf"
 
 # Similarly, some are only suitable for cross toolchains.
 # Remove these if host=target.
diff -Nbaur --exclude=*.info* binutils-2.16.1/gas/config/tc-arm.c binutils-2.16.1-arm/gas/config/tc-arm.c
--- binutils-2.16.1/gas/config/tc-arm.c	Tue May 17 20:51:33 2005
+++ binutils-2.16.1-arm/gas/config/tc-arm.c	Fri Sep 23 03:58:13 2005
@@ -12697,13 +12697,18 @@
   {"arm926ej-s",	ARM_ARCH_V5TEJ,	 FPU_ARCH_VFP_V2},
   {"arm946e-r0",	ARM_ARCH_V5TExP, FPU_ARCH_VFP_V2},
   {"arm946e",		ARM_ARCH_V5TE,   FPU_ARCH_VFP_V2},
+  {"arm946e-s",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
   {"arm966e-r0",	ARM_ARCH_V5TExP, FPU_ARCH_VFP_V2},
   {"arm966e",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
+  {"arm966e-s",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
+  {"arm968e-s",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
   {"arm10t",		ARM_ARCH_V5T,	 FPU_ARCH_VFP_V1},
+  {"arm10tdmi",		ARM_ARCH_V5T,	 FPU_ARCH_VFP_V1},
   {"arm10e",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
   {"arm1020",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
   {"arm1020t",		ARM_ARCH_V5T,	 FPU_ARCH_VFP_V1},
   {"arm1020e",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
+  {"arm1022e",		ARM_ARCH_V5TE,	 FPU_ARCH_VFP_V2},
   {"arm1026ejs",	ARM_ARCH_V5TEJ,	 FPU_ARCH_VFP_V2},
   {"arm1026ej-s",	ARM_ARCH_V5TEJ,	 FPU_ARCH_VFP_V2},
   {"arm1136js",		ARM_ARCH_V6,     FPU_NONE},
