diff -NBaur gcc-7.2.0/gcc/config/aarch64/aarch64-elf-raw.h gcc-7.2.0-aarch64/gcc/config/aarch64/aarch64-elf-raw.h
--- gcc-7.2.0/gcc/config/aarch64/aarch64-elf-raw.h	2017-01-01 12:07:43.905435000 +0000
+++ gcc-7.2.0-aarch64/gcc/config/aarch64/aarch64-elf-raw.h	2017-09-11 09:44:35.042230491 +0100
@@ -22,6 +22,7 @@
 #ifndef GCC_AARCH64_ELF_RAW_H
 #define GCC_AARCH64_ELF_RAW_H
 
+#define LINK_GCC_C_SEQUENCE_SPEC "--start-group %G %L %(libgloss) --end-group"
 #define STARTFILE_SPEC " crti%O%s crtbegin%O%s crt0%O%s"
 #define ENDFILE_SPEC \
   " crtend%O%s crtn%O%s " \
diff -NBaur gcc-7.2.0/gcc/config/aarch64/t-aarch64 gcc-7.2.0-aarch64/gcc/config/aarch64/t-aarch64
--- gcc-7.2.0/gcc/config/aarch64/t-aarch64	2017-01-01 12:07:43.905435000 +0000
+++ gcc-7.2.0-aarch64/gcc/config/aarch64/t-aarch64	2017-09-11 09:44:35.042230491 +0100
@@ -68,5 +68,7 @@
 		$(srcdir)/config/aarch64/cortex-a57-fma-steering.c
 
 comma=,
-MULTILIB_OPTIONS    = $(subst $(comma),/, $(patsubst %, mabi=%, $(subst $(comma),$(comma)mabi=,$(TM_MULTILIB_CONFIG))))
-MULTILIB_DIRNAMES   = $(subst $(comma), ,$(TM_MULTILIB_CONFIG))
+MULTILIB_OPTIONS    = mcmodel=large
+MULTILIB_DIRNAMES   = large
+MULTILIB_REQUIRED   = mcmodel=large
+MULTILIB_MATCHES    =
diff -NBaur gcc-7.2.0/gcc/gcc.c gcc-7.2.0-aarch64/gcc/gcc.c
--- gcc-7.2.0/gcc/gcc.c	2017-06-22 12:38:56.224791000 +0100
+++ gcc-7.2.0-aarch64/gcc/gcc.c	2017-09-11 09:44:35.042230491 +0100
@@ -779,6 +779,11 @@
 #endif
 #endif
 
+#ifndef LIBGLOSS_SPEC
+# define LIBGLOSS_SPEC "-lsysbase"
+#endif
+
+
 /* config.h can define STARTFILE_SPEC to override the default crt0 files.  */
 #ifndef STARTFILE_SPEC
 #define STARTFILE_SPEC  \
@@ -1074,6 +1079,7 @@
 static const char *lib_spec = LIB_SPEC;
 static const char *link_gomp_spec = "";
 static const char *libgcc_spec = LIBGCC_SPEC;
+static const char *libgloss_spec = LIBGLOSS_SPEC;
 static const char *endfile_spec = ENDFILE_SPEC;
 static const char *startfile_spec = STARTFILE_SPEC;
 static const char *linker_name_spec = LINKER_NAME;
@@ -1579,6 +1585,7 @@
   INIT_STATIC_SPEC ("lib",			&lib_spec),
   INIT_STATIC_SPEC ("link_gomp",		&link_gomp_spec),
   INIT_STATIC_SPEC ("libgcc",			&libgcc_spec),
+  INIT_STATIC_SPEC ("libgloss",			&libgloss_spec),
   INIT_STATIC_SPEC ("startfile",		&startfile_spec),
   INIT_STATIC_SPEC ("cross_compile",		&cross_compile),
   INIT_STATIC_SPEC ("version",			&compiler_version),
diff -NBaur gcc-7.2.0/libcpp/Makefile.in gcc-7.2.0-aarch64/libcpp/Makefile.in
--- gcc-7.2.0/libcpp/Makefile.in	2017-08-14 09:03:31.162084275 +0100
+++ gcc-7.2.0-aarch64/libcpp/Makefile.in	2017-09-11 09:44:35.042230491 +0100
@@ -208,8 +208,7 @@
 # Note that we put the dependencies into a .Tpo file, then move them
 # into place if the compile succeeds.  We need this because gcc does
 # not atomically write the dependency output file.
-COMPILE = $(COMPILE.base) -o $@ -MT $@ -MMD -MP -MF $(DEPDIR)/$*.Tpo
-POSTCOMPILE = @mv $(DEPDIR)/$*.Tpo $(DEPDIR)/$*.Po
+COMPILE = $(COMPILE.base) -o $@
 else
 COMPILE = source='$<' object='$@' libtool=no DEPDIR=$(DEPDIR) $(DEPMODE) \
 	  $(depcomp) $(COMPILE.base)
diff -NBaur gcc-7.2.0/libgcc/crtstuff.c gcc-7.2.0-aarch64/libgcc/crtstuff.c
--- gcc-7.2.0/libgcc/crtstuff.c	2017-01-01 12:07:43.905435000 +0000
+++ gcc-7.2.0-aarch64/libgcc/crtstuff.c	2017-09-11 09:44:35.046230483 +0100
@@ -47,6 +47,7 @@
 
 /* Target machine header files require this define. */
 #define IN_LIBGCC2
+#define USED_FOR_TARGET
 
 /* FIXME: Including auto-host is incorrect, but until we have
    identified the set of defines that need to go into auto-target.h,
diff -NBaur gcc-7.2.0/libgcc/Makefile.in gcc-7.2.0-aarch64/libgcc/Makefile.in
--- gcc-7.2.0/libgcc/Makefile.in	2017-04-19 09:08:44.446150000 +0100
+++ gcc-7.2.0-aarch64/libgcc/Makefile.in	2017-09-11 09:44:35.046230483 +0100
@@ -844,7 +844,7 @@
 # libgcc_eh.a, only LIB2ADDEH matters.  If we do, only LIB2ADDEHSTATIC and
 # LIB2ADDEHSHARED matter.  (Usually all three are identical.)
 
-c_flags := -fexceptions
+c_flags := -fno-exceptions
 
 ifeq ($(enable_shared),yes)
 
